### ISSUE TITLE
LSP4J is entangled with IDE

### ISSUE THREAD

**Issue by tivervac** [eclipse/xtext-eclipse#711]
Currently ide.server extensively uses LSP4J which makes it hard to get dependencies right.
I will be using [this example PR](https://github.com/eclipse/xtext-eclipse/pull/710) as basis for the explanation.

I wanted to create an adapter from `IOccurrenceComputer` to `IDocumentHighlightService` so that I only have to implement it once for it to work on LSP and Eclipse. The code is quite simple, just transform the inputs and the outputs.

The problem is that the `IDocumentHighlightService` API uses LSP4J arguments (`TextDocumentPositionParams` as input and `DocumentHighlightKind` as output) which requires to add LSP4J as a dependency to `xtext.ui`.

Another solution would be to do the transformations on the `xtext.ide` side, but that would require a dependency from `xtext.ide` to `xtext.ui` which is impossible.

In conclusion, there is no way to write an adapter from `xtext.ui` to `xtext.ide` without adding LSP4J as a dependency for `xtext.ui`. Creating new interfaces for the LSP4J arguments in `xtext.ide` and delegating those to said LSP4J objects could alleviate the problem.

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext-eclipse/issues/639

@svenefftinge @spoenemann we should investigate how we cann have lsp and eclipse independent ide parts so that they can be used from both. unfortunately some of the "bad" ide interfaces already use lsp4j in their api so we would need to replace that or build adapters or have a look if clients use the old interfaces and then call them.

it would be really nice if you could participate in finding a proper solution

**Comment by svenefftinge** [comment]
Why can't you just add LSP4J? Including LSP4J in IDE was a deliberate decision as it is a relatively small library without any special platform requirements.

**Comment by cdietrich** [comment]
for me it was a explicit decision not to add it to eclipse as explicit dependency otherwise the lsp4j dependency would have had to be added as required dependency in MANIFEST.MF.

**Comment by tivervac** [comment]
@cdietrich Why did you avoid that? To keep plugin bloat to a minimum?

**Comment by cdietrich** [comment]
i dont know and @spoenemann is not reachable (holiday)

**Comment by spoenemann** [comment]
We should just include LSP4J as a dependency of Xtext-Eclipse (xtext.ui)

**Comment by cdietrich** [comment]
but this implicates that lsp4j is part of simrel which no longer seems to be planned

**Comment by spoenemann** [comment]
See eclipse/lsp4j#207.

**Comment by cdietrich** [comment]
i want to postpone this to 2.16.
i want this to be tought through

**Comment by cdietrich** [comment]
we discussed this issue on our side and we really dislike the option to use lsp4j inside our eclipse api. thus i would prefer to have a lsp4j independent api in ide and adapt to that from lsp4j impl in xtext

**Comment by tivervac** [comment]
@cdietrich Not choosing any sides, but could you please elaborate on that decision?

**Comment by lievenlemiengre** [comment]
> we have one api and two adapters. eclipse adapter does not depend on lsp4j
=> we dont need to clutter our eclipse apis with lsp4j

I think this is the right way to go.

**Comment by svenefftinge** [comment]
Fine with me, too. Would just be nice if it doesn't break too much stuff downstream.

**Comment by dhuebner** [comment]
I already implemented a lot of `ide` services and hope it will still work with the new Xtext releases

**Comment by cdietrich** [comment]
@dhuebner can you give examples of what you have implemented. which api do you have overriden?

**Comment by dhuebner** [comment]
@cdietrich I think it is easier to tell which I didn't implemented yet. `IDocumentHighlightService` is one of services I didn't touch so far.
However I'm interested in discussions about possible backward compatibility solutions for LSP4j. Overall, this issue fills like an `epic` issue to me.

**Comment by cdietrich** [comment]
one point i really dislike about using lsp4j services in eclipse are api changes. they are very very common in lsp4j but very rare in eclipse apis.

**Comment by svenefftinge** [comment]
The reasons why I initially decided to not have another indirection between the Xtext user API and LSP4J:
 - LSP4J is a thin lib
 - LSP is well specified and documented
 - LSP protocol can be enhanced, e.g. I could add additional properties to the messages if the client understands them. Passing those things through indirection layers is much harder.

So it would be good to at least allow binding directly to the LSP abstraction level. It's ok if you want to introduce additional abstraction layers which are independent of it, but they should go on top.

**Comment by cdietrich** [comment]
@svenefftinge i dont get a commitment from your side to do the maintenance work and the api change work. if we have a lsp4j indenpendent api we dont have a problem.

**Comment by svenefftinge** [comment]
Did you actually read my comment?
I am not interested in a lsp4j-independent API, because it adds bloat for no good reasons.
But we have a need for a direct LSP-level abstraction (see above). And we have been implementing and maintaining it.

**Comment by cdietrich** [comment]
no. all the changes with either types are actually breaking changes which would strictly require a Xtext 3.0.
i can live with that in the lsp api part but i dont want that practice to sneak into eclipse apis as well.
